src/adaptors/mod.rs

```diff
@@ -761,6 +761,26 @@ macro_rules! impl_tuple_combination {
                 let n = self.iter.count();
                 checked_binomial(n, K).unwrap() + self.c.count()
             }
+
+            fn fold<B, F>(self, mut init: B, mut f: F) -> B
+            where
+                F: FnMut(B, Self::Item) -> B,
+            {
+                let Self { c, item, mut iter } = self;
+                init = c
+                    .map(|($($X),*,)| {
+                        let z = item.clone().unwrap();
```

Comment:
Would it make sense / improve performance to `unwrap` only once, possibly by doing the first iteration explicitly, then saying `let z = item.as_ref().unwrap()`, and `z.clone()` within the loop?

Reply:
It seemed like a good idea, but a test panicked on `unwrap` if done outside of `map`. I guess `item` might be none when `c` is empty (and therefore `unwrap` inside `map` is not called).

Reply:
@phimuemue Well I could avoid `unwrap` entirely (it passes tests as expected):

    if let Some(z) = item.as_ref() {
        init = c
            .map(|($($X),*,)| {
                (z.clone(), $($X),*)
            })
            .fold(init, &mut f);
    }